[content-list] Route for listing, filtering, sorting content.

clintandrewhall opened the pull request:
## Summary

This PR introduces a new internal API endpoint `POST /internal/content_management/list` that provides advanced listing capabilities for saved objects with rich filtering, sorting, and user enrichment features.

> [!NOTE]
> This is a pre-requisite for the new `ContentList` components, intended to replace `TableListView`.

### Features

- **Text search**: Full-text search across `title` and `description` fields using Elasticsearch `simple_query_string`
- **Tag filtering**: Include or exclude objects by tag references using nested queries
- **Favorites filtering**: Filter to only the current user's favorited objects
- **Creator filtering**: Filter by creator user profile UID, username, or email (with automatic resolution)
- **Flexible sorting**: Sort by `title`, `updatedAt`, `createdAt`, `createdBy`, `updatedBy`, `managed`, or custom attributes via runtime mappings
- **Pagination**: Standard offset-based pagination with configurable page size
- **User enrichment**: Response items include `createdByUser` and `updatedByUser` profile information (username, email, full name, avatar)

### Architecture

The route implementation is modularized into focused sub-modules:

```
routes/list/
├── index.ts           # Barrel exports
├── route.ts           # Route registration and handler
├── types.ts           # TypeScript interfaces
├── query_builder.ts   # ES query, runtime mappings, and sort construction
├── transformers.ts    # ES hit transformation and user enrichment
└── user_resolution.ts # Username/email to UID resolution and profile fetching
```

### Security

- Authorization is delegated to the Saved Objects Client (users only see objects they have access to)
- Field name validation prevents script injection in runtime mappings
- Creator filter resolution uses exact matching only (no partial/fuzzy matching)
- Distinct creators aggregation respects query scope for security

### Testing

Comprehensive unit tests added for all modules with ~77% line coverage:
- `route.test.ts` - Route registration and handler integration tests
- `query_builder.test.ts` - Query building utility tests
- `transformers.test.ts` - Transformation function tests
- `user_resolution.test.ts` - User profile resolution tests

### Documentation

All exported functions and types include JSDoc documentation with:
- Detailed descriptions
- `@param` and `@returns` tags
- `@example` usage blocks


> [!NOTE]
> This PR was written in Cursor with the assistance of `claude-4.5-opus-high`

clintandrewhall committed:
[content-list] Route for listing, filtering, sorting content.

This PR introduces a new internal API endpoint `POST /internal/content_management/list` that provides advanced listing capabilities for saved objects with rich filtering, sorting, and user enrichment features.

- **Text search**: Full-text search across `title` and `description` fields using Elasticsearch `simple_query_string`
- **Tag filtering**: Include or exclude objects by tag references using nested queries
- **Favorites filtering**: Filter to only the current user's favorited objects
- **Creator filtering**: Filter by creator user profile UID, username, or email (with automatic resolution)
- **Flexible sorting**: Sort by `title`, `updatedAt`, `createdAt`, `createdBy`, `updatedBy`, `managed`, or custom attributes via runtime mappings
- **Pagination**: Standard offset-based pagination with configurable page size
- **User enrichment**: Response items include `createdByUser` and `updatedByUser` profile information (username, email, full name, avatar)

The route implementation is modularized into focused sub-modules:

```
routes/list/
├── index.ts           # Barrel exports
├── route.ts           # Route registration and handler
├── types.ts           # TypeScript interfaces
├── query_builder.ts   # ES query, runtime mappings, and sort construction
├── transformers.ts    # ES hit transformation and user enrichment
└── user_resolution.ts # Username/email to UID resolution and profile fetching
```

- Authorization is delegated to the Saved Objects Client (users only see objects they have access to)
- Field name validation prevents script injection in runtime mappings
- Creator filter resolution uses exact matching only (no partial/fuzzy matching)
- Distinct creators aggregation respects query scope for security

Comprehensive unit tests added for all modules with ~77% line coverage:
- `route.test.ts` - Route registration and handler integration tests
- `query_builder.test.ts` - Query building utility tests
- `transformers.test.ts` - Transformation function tests
- `user_resolution.test.ts` - User profile resolution tests

All exported functions and types include JSDoc documentation with:
- Detailed descriptions
- `@param` and `@returns` tags
- `@example` usage blocks

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/375255)
* Commit: 7bd2926ecc4b22c0f17fc2e24c33c9956a51b9ea

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/375255#019b3359-7090-4af9-a38b-e67906c113f9)

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/375255","number":375255,"commit":"7bd2926ecc4b22c0f17fc2e24c33c9956a51b9ea"}],"number":375255}
buildkite-pr-comment-->

kibanamachine committed:
Changes from node scripts/eslint_all_files --no-cache --fix

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/375276)
* Commit: 2255d73de8e85dbfd6451f46efc4f74e8fee8144

### Failed CI Steps
* [Linting](https://buildkite.com/elastic/kibana-pull-request/builds/375276#019b3393-b71e-451b-86f5-594cbd678f0f)

### History
* :broken_heart: [Build #375270](https://buildkite.com/elastic/kibana-pull-request/builds/375270) failed 9d60eab5815984125d71c255dbd255c28bc722c6
* :broken_heart: [Build #375255](https://buildkite.com/elastic/kibana-pull-request/builds/375255) failed 7bd2926ecc4b22c0f17fc2e24c33c9956a51b9ea

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/375276","number":375276,"commit":"2255d73de8e85dbfd6451f46efc4f74e8fee8144"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/375270","number":375270,"commit":"9d60eab5815984125d71c255dbd255c28bc722c6"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/375255","number":375255,"commit":"7bd2926ecc4b22c0f17fc2e24c33c9956a51b9ea"}],"number":375276}
buildkite-pr-comment-->

clintandrewhall committed:
Update to fix eslint

Dosant commented on src/platform/plugins/shared/content_management/server/routes/list/query_builder.ts:
```diff
@@ -0,0 +1,328 @@
+  const rootFields = ['updatedAt', 'createdAt', 'updatedBy', 'createdBy', 'managed'];
+
+  const mappings: estypes.MappingRuntimeFields = {
+    sortable_title: {
```
Could it be that the better way to approach this is to make "title" sortable in all supported saved objects?

Also, does this runtime approach plays nicely if title is already sortable? I think you mentioned that "maps" has it sortable already

clintandrewhall replied:
Opposite-- maps doesn't have the keyword mapping.

I'll see what logic I can add.

clintandrewhall replied:
Looking into this, I found the following:

> 1. Most saved object types use `title: { type: 'text' }` without keyword (dashboard, maps, search, visualization, etc.)
> 2. The runtime field approach works uniformly for all types
> 3. The performance impact of runtime fields is minimal for listing operations

The logic to determine if the runtime field is necessary would vary by object type, and require interrogating a number of mappings.

Short of hard-coding _exceptions_ to needing the field, I'm inclined to agree that this can just be the de-facto approach.

Your thoughts?

Dosant commented on src/platform/plugins/shared/content_management/server/routes/list/query_builder.ts:
```diff
@@ -0,0 +1,328 @@
+  const rootFieldMap: Record<string, string> = {
+    updatedAt: 'updated_at',
+    createdAt: 'created_at',
+    updatedBy: 'updated_by',
```
I remembered why created_by / updated_by weren't sortable in the UI. this is because we store user id and it doesn't make sense to sort by user id, we should be sorting by user-facing naming. 

Is this handled somehow? Otherwise we should keep this unsortable

clintandrewhall replied:
Thanks for that insight!  I can adjust, since I'm gathering the metadata for the response!

Dosant commented on src/platform/plugins/shared/content_management/server/routes/list/route.ts:
```diff
@@ -0,0 +1,377 @@
+          }
+
+          // Fetch user profiles and enrich items.
+          const userInfoMap = await fetchUserProfiles(Array.from(userIds), coreStart, logger);
```
if we fetch 100 dashboards and there is 1 creator - we return the same user profile 100x 

Maybe we should keep it as today where it is enough to return profile_id and client-side we have smart components with react query that know how to render profiles from user cache <UserAvatar id={id} />  ?

or at least return profiles as separate map, but I'd prefer not include them at all

Dosant replied:
`tag:(clint.hall@elastic.co)` 

I think this is a cool feature, but since it's new, we might need more consideration before merging it in production. There are definitely some edge cases we should think through. 

bulkGet might be a bottleneck when fetching 50+ users: 

https://github.com/elastic/kibana/blob/8773c7482f5ba5b245724ff351cc5ab83b558dd0/x-pack/platform/plugins/shared/security/server/user_profile/user_profile_service.ts#L257-L281

Dosant reviewed (commented):
Looks exciting. I mentioned some concerns that I have. I also think that in the end this should be reviewed and approved by the core team, because there is too much knowledge about saved objects lingering here.

Dosant commented on src/platform/plugins/shared/content_management/server/routes/list/route.ts:
```diff
@@ -0,0 +1,377 @@
+          if (hasCreatedByFilter) {
+            const hasNonUidValues = createdBy.some((v) => !isUserProfileUid(v));
+
+            if (hasNonUidValues) {
```
If we implement this, I think we should use suggestProfiles instead. But currently, this API is high license only. So this could be either a high license feature (which I think is fine) or we can work with the security team to make this API route available without a license.

clintandrewhall replied:
Great idea.  I went this way to avoid the license check.  Not sure how we'd indicate that a license is preventing user profile filtering in the UX.

Dosant replied:
This is why the `created_by` filter wasn't supported in the query string - we couldn't filter by name without mapping it to a user ID.

For now, we could keep filtering only by user ID. For full support to filter by name/email, we should rediscuss the license requirements with the security teams.

elena-shostak replied:
For the suggest we need at least standard license ([allowUserProfileCollaboration](https://github.com/elastic/kibana/blob/main/x-pack/platform/plugins/shared/security/common/licensing/license_service.ts#L130)). I don't see any reasons not allowing it on basic, but we need @bytebilly for this

clintandrewhall committed:
optimize list endpoint with favorites utility, user caching, response restructure

- Extract `getFavoritesForTypes` to `@kbn/content-management-favorites-server` for reusable favorites lookup
- Add LRU cache for user profiles (1000 entries, 5min TTL) to reduce API calls
- Move user info from items to top-level `users` map to avoid duplication
- Fix tag filtering to use OR logic (match ANY tag instead of ALL)

clintandrewhall commented on src/platform/plugins/shared/content_management/server/routes/list/route.ts:
```diff
@@ -0,0 +1,377 @@
+              return undefined;
+            }
+
+            const favoritesSoClient = coreContext.savedObjects.getClient({
```
Addressed.

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/375799)
* Commit: 54a886dc8432474b9593eeaa3039a2749dd925d3

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/375799#019b3891-2e5e-4a15-aeb0-331135cee187)

### History
* :broken_heart: [Build #375276](https://buildkite.com/elastic/kibana-pull-request/builds/375276) failed 2255d73de8e85dbfd6451f46efc4f74e8fee8144
* :broken_heart: [Build #375270](https://buildkite.com/elastic/kibana-pull-request/builds/375270) failed 9d60eab5815984125d71c255dbd255c28bc722c6
* :broken_heart: [Build #375255](https://buildkite.com/elastic/kibana-pull-request/builds/375255) failed 7bd2926ecc4b22c0f17fc2e24c33c9956a51b9ea

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/375799","number":375799,"commit":"54a886dc8432474b9593eeaa3039a2749dd925d3"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/375276","number":375276,"commit":"2255d73de8e85dbfd6451f46efc4f74e8fee8144"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/375270","number":375270,"commit":"9d60eab5815984125d71c255dbd255c28bc722c6"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/375255","number":375255,"commit":"7bd2926ecc4b22c0f17fc2e24c33c9956a51b9ea"}],"number":375799}
buildkite-pr-comment-->

kibanamachine committed:
Changes from node scripts/eslint_all_files --no-cache --fix

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/375806)
* Commit: ae279f60569445f8726391d45e2369787de3d178

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/375806#019b38a7-589c-495d-9419-b9e0d1e2c19a)

### History
* :broken_heart: [Build #375799](https://buildkite.com/elastic/kibana-pull-request/builds/375799) failed 54a886dc8432474b9593eeaa3039a2749dd925d3
* :broken_heart: [Build #375276](https://buildkite.com/elastic/kibana-pull-request/builds/375276) failed 2255d73de8e85dbfd6451f46efc4f74e8fee8144
* :broken_heart: [Build #375270](https://buildkite.com/elastic/kibana-pull-request/builds/375270) failed 9d60eab5815984125d71c255dbd255c28bc722c6
* :broken_heart: [Build #375255](https://buildkite.com/elastic/kibana-pull-request/builds/375255) failed 7bd2926ecc4b22c0f17fc2e24c33c9956a51b9ea

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/375806","number":375806,"commit":"ae279f60569445f8726391d45e2369787de3d178"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/375799","number":375799,"commit":"54a886dc8432474b9593eeaa3039a2749dd925d3"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/375276","number":375276,"commit":"2255d73de8e85dbfd6451f46efc4f74e8fee8144"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/375270","number":375270,"commit":"9d60eab5815984125d71c255dbd255c28bc722c6"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/375255","number":375255,"commit":"7bd2926ecc4b22c0f17fc2e24c33c9956a51b9ea"}],"number":375806}
buildkite-pr-comment-->

Dosant reviewed (commented):
I'll summarize my remaining concerns in a single comment to make it easier to navigate: 


1. When allowing "unknown" attributes we're circumventing bwc layer and if "unknown" attributes are used on UI we're hitting problems we had we with client side so client. 

Yes, it's smoother DX and more possibilities, but it also doesn't follow the current guidelines from core. So we should explicitly agree the approach is fine or if we address somehow. 


2. userprofile.bulkGet worst case for filtering should be too slow now. The cache mitigates it, but if we keep it, it should be in the user profile service in my opinion, not as a quick patch here.

Instead I'd suggest we use suggestUser and drop username/email feature where suggestUser not available and long-term enable it when/if it becomes available

3. Favorites part imo is fine as a tech debt, but at least building a needed so client should also be moved to favorites package
